src/Microsoft.Azure.WebJobs.Host/Scale/DynamicTargetValueProvider.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+    internal class DynamicTargetValueProvider : IDynamicTargetValueProvider
```

Comment:
Extensions using ConcurrencyManager already have access to the persisted concurrency value via their calls to ConcurrencyManager.GetStatus which they make as part of dynamic scaling. There's a hosted service running [ConcurrencyManagerService](https://github.com/Azure/azure-webjobs-sdk/blob/62bbc94767672b17600f8329a5d36fee364a7c72/src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyManagerService.cs#L65) responsible for reading the status and applying it on startup, as well as persisting updates on a timer. So I'm wondering if rather than adding a new service to read this value, we can't instead just use the value provided by [ConcurrencyStatus.CurrentConcurrency](https://github.com/Azure/azure-webjobs-sdk/blob/1f7f383c2f48237a76f9b3112c79e4a3800ec4d9/src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyStatus.cs#L71)?

Reply:
@chiangvincent, as @mathewc said above we can read the dynamic concurrency value the function host is currently using so there is no need to read from the blob. Do you think we can go this path?

Scale controller has the logic to cache the concurrency value, also it can be expired and fallback value will be returned in this case.

Reply:
Interesting, this should be doable in scale controller, as long as there are no permissions issues for the blob in storage when using the ConcurrencyManagerService.

Reply:
I do not think it would be doable in SC as `ConcurrencyManagerService `have access to executions and dynamic concurrency engine directly to return the dynamic concurrency value. `ConcurrencyManagerService `reads the blob once on startup to set up initial concurrency and then primary host instance stores current dynamic value to the blob.

Reply:
Question is if it's ok to use `ConcurrencyManager.GetStatus` value for target base scale. The value is calculated dynamically based on executions and different metrics.

Reply:
@pragnagopa @TsuyoshiUshio as fyi

Reply:
The current logic in scale controller has an additional layer of logic on retrieval of this value. For example, logic to check if the value has stabilized and if we should use it at all. As long as we have stabilization logic built on top of the value retrieved from `ConcurrencyManager.GetStatus` it should be ok to use.

Reply:
so it looks like we still need `IDynamicTargetValueProvider` to have the stabilization logic.

Reply:
The concurrency value for each function is adjusted over time. Initially it starts at 1 and over time gets adjusted based on extensions calling into ConcurrencyManager.GetStatus. Periodically that value is written to storage by ConcurrencyManagerService [here](https://github.com/Azure/azure-webjobs-sdk/blob/62bbc94767672b17600f8329a5d36fee364a7c72/src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyManagerService.cs#L111). So what do you mean exactly by "unstable" - it's dynamic, so yes will tend to stabilize over time.

My main point was that we don't need a new service that extensions use to read this value from storage, since extensions already have access to the source of that value already - ConcurrencyStatus.CurrentConcurrency. An extension when asked to determine it's scale status (vote/TargetWorkerCount) should just use the most current information it has.

Reply:
Regarding main point it makes sense - we do not need to read the value from storage as we can get current value using `ConcurrencyManager.GetStatus`. Agree.

"Unstable" means that the dynamic value can fluctuate a lot and we want to ignore peak values to avoid overscale/underscale. @chiangvincent correct me if I am wrong. Question is do we want this to be implemented as in SC controller or we want to rely only on current dynamic value.

Reply:
I recommend going with what is built into SDK - current dynamic value. We can have thresholds in SC specifically to keep checks on overscale / underscale